### ISSUE TITLE
Inter-operate with platform-specific "favorite directories"

### ISSUE THREAD

**Issue by charlesdaniels** [fyne-io/fyne#1238]
We should consider adding some kind of capability, possible as part of the driver, that exposes the platform's "favorite directories". For a start, we could just read them and add to the side-bar, but in a perfect world we would interact with these bi-bidirectionally.

![windows_favorites](https://user-images.githubusercontent.com/11738895/89669781-0a7b8e00-d8ae-11ea-9b2c-4755be2233bb.png)

![macos_favorites](https://user-images.githubusercontent.com/11738895/89669670-def8a380-d8ad-11ea-9329-6500984231d6.png)

**Comment by noerw** [comment]
These files contain the bookmarks on different systems:
- Linux Gnome (nautlius): `~/.config/gtk-3.0/bookmarks`: line separated `file://` or `dav://` URIs
- Linux KDE (dolphin): [`~/.local/share/apps/kfileplaces/bookmarks.xml`](https://userbase.kde.org/Dolphin/File_Management#Bookmarks_and_Places)
- Windows 10: `%userprofile%\AppData\Roaming\Microsoft\Windows\Recent\CustomDestinations`
- OSX: [`~/Library/Preferences/com.apple.finder.plist`](https://georgegarside.com/blog/macos/finder-sidebar-preferences-location/)

**Comment by andydotxyz** [comment]
This is going to be a lot of work especially as we consider all the different linux desktops and distributions.

**Comment by charlesdaniels** [comment]
My $0.02 -- we should build a library that solves this on it's own, and then use it as an internal dependency of Fyne. People building other programs could well want that as an independent thing from Fyne also.

**Comment by noerw** [comment]
In the meantime i found this; which basically implements what @charlesdaniels described:
https://github.com/adrg/xdg
On linux, the XDG-Desktop spec is followed.

**Comment by andydotxyz** [comment]
As far as I know XDG is only followed on Linux.  None of Windows, macOS, iOS or Android use this spec.

Also, as you noted before, the user-configured list of directories (which this ticket aims to match) is not part of the XDG but  instead specific to the desktop shell that the user prefers.
With regards to which location should be used for each XDG known location we have added support for that in 1.4 already, having been tracked in #1248.

**Comment by AlbinoGeek** [comment]
Linux **global**: [FreeDesktop Bookmark Specification](https://www.freedesktop.org/wiki/Specifications/desktop-bookmark-spec/)

I can also put together a list of known XDG formats for this if it will help this along, having implemented favourites in my app already.

**Comment by Lewiscowles1986** [comment]
The files in the CustomDestinations folder on windows follow the [Structured Storage format by Microsoft](https://docs.microsoft.com/en-us/windows/win32/stg/structured-storage-start-page); it reads (from their docs) like someone on a night out designed the format while unable to function.

**Comment by Lewiscowles1986** [comment]
Going to see if I can use some off the shelf reference code to read (C#) https://github.com/ironfede/openmcdf

If I can find a way to read in this and the C/CPP microsoft examples; it might become more obvious what they are doing and why. Frustratingly guidance on [This stackoverflow on the files](https://stackoverflow.com/questions/4429944/decrypt-read-modify-automaticdestinations-ms-and-or-customdestinations-ms) suggests the format has changed;

**Comment by Lewiscowles1986** [comment]
Interestingly I can enumerate all the values in powershell using the following

```
$QuickAccess = New-Object -ComObject shell.application
$QuickAccess.Namespace("shell:::{679f85cb-0220-4080-b29b-5540cc05aab6}").Items()
```

I need to find out where the heck `{679f85cb-0220-4080-b29b-5540cc05aab6}` exists as regedit cannot find it; and neither can find.

**Comment by AlbinoGeek** [comment]
Honest question, does it matter? Powershell is available in every supported
version of windows, why not just fetch it through powershell?

**Comment by Lewiscowles1986** [comment]
Do you mean sub-process, and then interpret the output?

**Comment by AlbinoGeek** [comment]
No, literally speak to the powershell host inside golang, there are
multiple packages to do this with, here is an example:

https://github.com/KnicKnic/go-powershell

**Comment by AlbinoGeek** [comment]
That particular package is probably overkill for what you want to do here,
execute a single command and get it output, but you could look at that code
as a good example for how they did that, and just use that one routine.

**Comment by Lewiscowles1986** [comment]
If you would like to chase that path, feel free. I'm trying to extract the win32 API calls. It'll just make me feel better.

**Comment by Lewiscowles1986** [comment]
`Source file: %userprofile%\AppData\Roaming\Microsoft\Windows\Recent\AutomaticDestinations\f01b4d95cf55d32a.automaticDestinations-ms` contained all these entries for me. Curious it should be in a single file (not one other file had any destinations for me).

* [This medium article led me to a tool](https://nasbench.medium.com/windows-forensics-analysis-windows-artifacts-part-ii-71b8fa68d8a1)
* [JLECmd](https://ericzimmerman.github.io/#:~:text=all%20the%20things-,JLECmd,-1.4.0.0)

In case it matters, I'm on Windows 10 64-bit home, Version	10.0.19044 Build 19044

I've still not found a way to ensure this is the right file or get this information in go.

**Comment by Lewiscowles1986** [comment]
Interestingly enough [this link about backing up those explorer pinned items](https://www.tenforums.com/tutorials/175934-how-backup-restore-quick-access-pinned-folders-windows-10-a.html) suggests that the same file as I found my pinned items in, should be backed up. Perhaps there is a standard list, or a canonical path to the file?

**Comment by Lewiscowles1986** [comment]
`Source file: %userprofile%\AppData\Roaming\Microsoft\Windows\Recent\AutomaticDestinations\f01b4d95cf55d32a.automaticDestinations-ms` does read with the `Structured Storage eXplorer`, which I had to compile from source of the [OpenMCDF GitHub project](https://github.com/ironfede/openmcdf)

**Comment by Lewiscowles1986** [comment]
https://github.com/EricZimmerman/OleCf/blob/master/OleCf/Header.cs is about reading this file-type. Which is apparently shared with old pre docx word documents.

https://github.com/richardlehane/mscfb/network is a go reader of the file-type; so something akin to https://github.com/EricZimmerman/JumpList should be used to parse out the locations.

I Think a benefit of this will also be that for windows, this software currently seems to assume ~/Documents will exist. I know from experience that I sometimes move this folder, and that Windows supports that; so this work could remove a classification of bug.

Here is a sample repo I made using richardlehane with EricZimmerman C# code as a reference I hacked it together https://github.com/Lewiscowles1986/read-pinned-folders
